core/templates/pages/exploration-editor-page/improvements-tab/improvements-tab.component.ts

```diff
@@ -13,10 +13,117 @@
+        // service.
+        $q.when(ExplorationImprovementsService.isImprovementsTabEnabledAsync())
+          .then(improvementsTabIsEnabled => {
+            if (!improvementsTabIsEnabled) {
+              return;
+            }
+
+            const expStats = (
+              ExplorationImprovementsTaskRegistryService.getExplorationStats());
+            this.completionRate = (
+              (expStats && expStats.numStarts > 0) ?
+                (expStats.numCompletions / expStats.numStarts) : 0);
+            this.completionRateAsPercent = (
+              Math.round(100.0 * this.completionRate) + '%');
+
+            const hbrTasks = (
+              ExplorationImprovementsTaskRegistryService
+                .getOpenHighBounceRateTasks());
+            const iflTasks = (
+              ExplorationImprovementsTaskRegistryService
+                .getOpenIneffectiveFeedbackLoopTasks());
+            const ngrTasks = (
+              ExplorationImprovementsTaskRegistryService
+                .getOpenNeedsGuidingResponsesTasks());
+            const siaTasks = (
+              ExplorationImprovementsTaskRegistryService
+                .getOpenSuccessiveIncorrectAnswersTasks());
+            this.getNumTasks = () => (
+              hbrTasks.length + iflTasks.length + ngrTasks.length +
+              siaTasks.length);
+            this.getNumExpLevelTasks = () => hbrTasks.length;
+            this.getNumConceptLevelTasks = () => iflTasks.length;
+            this.getNumCardLevelTasks = () => ngrTasks.length + siaTasks.length;
+            this.hasCriticalTasks = () => ngrTasks.length > 0;
+            this.getExplorationHealth = () => {
+              if (this.hasCriticalTasks()) {
+                return ImprovementsConstants.EXPLORATION_HEALTH_TYPE_CRITICAL;
+              } else if (this.getNumTasks() > 0) {
+                return ImprovementsConstants.EXPLORATION_HEALTH_TYPE_WARNING;
+              } else {
+                return ImprovementsConstants.EXPLORATION_HEALTH_TYPE_HEALTHY;
+              }
+            };
+
+            this.stateTasksObject = (
+              ExplorationImprovementsTaskRegistryService
+                .makeStateTasksObject());
+
+            const stateRetentions: Map<string, number> = new Map(
+              Object.keys(this.stateTasksObject).map(stateName => {
+                const {numCompletions, totalHitCount} = (
+                  expStats.getStateStats(stateName));
+                const retentionRate = (
+                  totalHitCount ? (numCompletions / totalHitCount) : 0);
+                return [stateName, Math.round(100.0 * retentionRate)];
+              }));
+            this.getStateRetention = (stateName: string) => {
+              return stateRetentions.get(stateName);
+            };
+            this.getStateNumCardLevelTasks = (stateName: string) => {
+              const {ngrTask, siaTask} = this.stateTasksObject[stateName];
```

Comment:
Is stateTasksObject a domain object? If not, it should be.

Reply:
Rather than assembling custom data structures to satisfy ng-repeat, I think I would suggest getting the list of states, having the ng-repeat be over the list of states, and calling getNgrTask(stateName) from the template.